### PR TITLE
feat: Pos invoice coupon code feature

### PR DESCRIPTION
**Pull request by nemesis189**
**Description:**
- This feature allows you to apply a coupon code created through the existing **Coupon Code** doctype at the POS checkout screen.
- Lets you type in the coupon code and a dropdown of relevant coupons will pop up.
 

**Steps:**
1. Create a pricing rule for 'Selling' such as the following:
Screenshot 2021-08-27 at 3 27 30 PM
Screenshot 2021-08-27 at 3 28 05 PM
**If the 'Apply On' field has item group selected, add a subsequent item group in the table, same for other options too.
The discount will be applicable only for those items which come under the applied option, here its Item Group - POS test items, so items with POS test items as their 'Item Group' only will have this applied.**

2. Create a Coupon Code
Screenshot 2021-08-27 at 3 28 05 PM

3. Select the created coupon code at the POS screen

`no-docs`

### CODE REVIEW

File: erpnext/accounts/doctype/pricing_rule/pricing_rule.py [hunk 1]
```diff
@@ -198,12 +198,19 @@ def apply_pricing_rule(args, doc=None):
 	set_serial_nos_based_on_fifo = frappe.db.get_single_value("Stock Settings",
 		"automatically_set_serial_nos_based_on_fifo")
 
+	item_code_list = tuple(item.get('item_code') for item in item_list)
+	query_items = frappe.get_all('Item', fields=['item_code','has_serial_no'], filters=[['item_code','in',item_code_list]],as_list=1)
+	serial_no_list = dict()
+	for i_c,val in query_items:
+		serial_no_list.setdefault(i_c,val)
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pricing_rule/pricing_rule.py]:
```suggestion
	serialized_items = dict()
	for item_code, val in query_items:
		serialized_items.setdefault(iitem_code, val)
```

File: erpnext/accounts/doctype/pricing_rule/pricing_rule.py [hunk 2]
```diff
@@ -198,12 +198,19 @@ def apply_pricing_rule(args, doc=None):
...
+		serial_no_list.setdefault(i_c,val) 
+	
 	for item in item_list:
 		args_copy = copy.deepcopy(args)
 		args_copy.update(item)
 		data = get_pricing_rule_for_item(args_copy, item.get('price_list_rate'), doc=doc)
 		out.append(data)
-		if not item.get("serial_no") and set_serial_nos_based_on_fifo and not args.get('is_return'):
+		
+		if serial_no_list.get(item.get('item_code')) and not item.get("serial_no") and set_serial_nos_based_on_fifo and not args.get('is_return'):
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pricing_rule/pricing_rule.py]:
```suggestion
		if serialized_items.get(item.get('item_code')) and not item.get("serial_no") and set_serial_nos_based_on_fifo and not args.get('is_return'):
```

**Review by nextchamp-saqib — DISMISSED** [review]

File: erpnext/accounts/doctype/pricing_rule/pricing_rule.py [hunk 3]
```diff
@@ -198,12 +198,19 @@ def apply_pricing_rule(args, doc=None):
 	set_serial_nos_based_on_fifo = frappe.db.get_single_value("Stock Settings",
 		"automatically_set_serial_nos_based_on_fifo")
 
+	item_code_list = tuple(item.get('item_code') for item in item_list)
+	query_items = frappe.get_all('Item', fields=['item_code','has_serial_no'], filters=[['item_code','in',item_code_list]],as_list=1)
+	serialized_items = dict()
+	for item_code, val in query_items:
+		serialized_items.setdefault(iitem_code, val)
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pricing_rule/pricing_rule.py:205]:
```suggestion
		serialized_items.setdefault(item_code, val)
```

**Review by nextchamp-saqib — DISMISSED** [review]

File: erpnext/accounts/doctype/pos_invoice/pos_invoice.json [hunk 4]
```diff
@@ -1550,6 +1551,15 @@
    "no_copy": 1,
    "options": "Sales Invoice",
    "read_only": 1
+  },
+  {
+   "depends_on": "coupon_code",
+   "fieldname": "coupon_code",
+   "fieldtype": "Link",
+   "hidden": 1,
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice/pos_invoice.json]:
```suggestion
```

**Review by nextchamp-saqib — APPROVED** [review]

**Comment by sagarvora** [comment]
Why hasn't the `modified` timestamp been updated?  Did you modify `pos_invoice.json` directly?

**Comment by nemesis189** [comment]
> Why hasn't the `modified` timestamp been updated? Did you modify `pos_invoice.json` directly?

No, I didn't touch the json file

**Comment by nextchamp-saqib** [comment]
> Why hasn't the `modified` timestamp been updated? Did you modify `pos_invoice.json` directly?

I gues due to this 086053dc5a91998af973e91c3ce60bcfc64b45e9, while resolving conflicts, modified timestamp is the latest one though. Will it cause a problem? @sagarvora

**Comment by sagarvora** [comment]
@nextchamp-saqib Yes, this will cause an issue. The DocType only gets synced with ERP if the modified timestamp is newer (not same or older).

**Pull request merged**